keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepositoryTests.kt

```diff
@@ -0,0 +1,344 @@
+package com.netflix.spinnaker.keel.persistence
+
+import com.netflix.spinnaker.keel.SPINNAKER_API_V1
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ResourceSpec
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+import com.netflix.spinnaker.keel.api.artifacts.DockerArtifact
+import com.netflix.spinnaker.keel.api.id
+import com.netflix.spinnaker.keel.core.api.SubmittedDeliveryConfig
+import com.netflix.spinnaker.keel.core.api.SubmittedEnvironment
+import com.netflix.spinnaker.keel.core.api.SubmittedResource
+import com.netflix.spinnaker.keel.core.api.normalize
+import com.netflix.spinnaker.keel.core.api.resources
+import com.netflix.spinnaker.keel.events.ArtifactRegisteredEvent
+import com.netflix.spinnaker.keel.events.ResourceCreated
+import com.netflix.spinnaker.keel.events.ResourceUpdated
+import com.netflix.spinnaker.keel.exceptions.DuplicateResourceIdException
+import com.netflix.spinnaker.keel.resources.ResourceTypeIdentifier
+import com.netflix.spinnaker.keel.test.DummyResourceSpec
+import com.netflix.spinnaker.keel.test.resource
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import io.mockk.mockk
+import io.mockk.verify
+import java.time.Clock
+import java.time.Duration
+import org.springframework.context.ApplicationEventPublisher
+import strikt.api.expect
+import strikt.api.expectCatching
+import strikt.api.expectThat
+import strikt.api.expectThrows
+import strikt.assertions.failed
+import strikt.assertions.first
+import strikt.assertions.hasSize
+import strikt.assertions.isA
+import strikt.assertions.isEmpty
+import strikt.assertions.isEqualTo
+import strikt.assertions.isTrue
+import strikt.assertions.succeeded
+
+abstract class CombinedRepositoryTests<D : DeliveryConfigRepository, R : ResourceRepository, A : ArtifactRepository> :
```

Comment:
Does it make sense to combine all these tests? I wonder if we could have kept them as separate test classes that each tested a particular aspect of the combined repo.

Reply:
I wanted to test both sql and non-sql for create & delete flows, so I thought it made sense to pull them both here. Would you prefer I split them out? If so, can you suggest what categories you'd like? I agree that small test files are easier to reason about.

Reply:
There actually aren't that many test in there, just create + verify that we start checking things + delete + validation. And, they all care about all three repositories. So I'm inclined to leave them in the same file for now, and split them up if they get unwieldy.

Reply:
Okay, maybe I misunderstood what I was seeing. It looked like you merged all the tests for the other repositories

Reply:
Oh no not at all. Those are all separate. This is just the muli-repo create + delete tests. Let me add a comment clarifying that.